### PR TITLE
Change the API we use for downloading configs from S3

### PR DESCRIPTION
**Pull request by jmhooper**
**Why**: To enable the effort for the IDP to download it's own configs from secrets and app-secrests

**How**:

1. Replace the `#s3` method with 2 separate methods, `#app_secrets_s3` and `#secrets_s3`.
2. Replace the `#download_configs` method with a `#download_config` that downloads 1 config at a time.
   Currently this is how we use it and likely how we'll use it going forward.
   This change helps this method to mirror the `#read_config` methodl
3. Add a `#read_config` method which reads an object form S3 and returns the body as a string.

### CODE REVIEW

**Comment by jmhooper** [comment]
I'll make sure to bump the version before I merge

File: lib/identity/hostdata/s3.rb [hunk 1]
```diff
@@ -16,22 +16,6 @@ def initialize(bucket:, env:, region:, s3_client: nil, logger: nil)
...
-
-      def s3_client
-        @s3_client ||= Aws::S3::Client.new(
-          region: region,
-          http_open_timeout: 5,
-          http_read_timeout: 5
-        )
-      end
-
       def download_config(s3_path, local_path)
```
Comment (zachmargolis) [inline lib/identity/hostdata/s3.rb:19]: Two ideas:
1. WDYT if we made these keyword arguments to be clearer
2. What about renaming to `download_file` since these could be configs, secrets, etc etc?

```suggestion
      def download_file(s3_path:, local_path:)
```

File: lib/identity/hostdata/s3.rb [hunk 2]
```diff
@@ -45,6 +29,29 @@ def download_config(s3_path, local_path)
           response_target: local_path
         )
       end
+
+      def read_config(s3_path)
+        key = format(s3_path, env: env).sub(%r|\A/|, '')
```
Comment (zachmargolis) [inline lib/identity/hostdata/s3.rb:34]: I'd like to find a way to clean up some duplication.

Part of me wants this to share an impl with the other, and call with nil `local_path`, but I think that would get a little hard to follow

As an alternative, can we at least extract a helper for the ley? maybe `build_key(s3_path)` so that if we ever add more keys to interpolate, we do that in one place?

**Review by zachmargolis — APPROVED** [review]
Please make sure to do a big VERSION bump as well! this is breaking changes

File: lib/identity/hostdata.rb [hunk 3]
```diff
@@ -81,16 +81,31 @@ def self.in_datacenter
...
         env: env,
         region: ec2.region,
         logger: logger,
         s3_client: s3_client,
-        bucket: "login-gov.app-secrets.#{ec2.account_id}-#{ec2.region}"
+        bucket: bucket
+      )
+    end
+
+    # @return [S3] An S3 object onfigured to use the secrets bucket
```
Comment (zachmargolis) [inline lib/identity/hostdata.rb:98]: typo

```suggestion
    # @return [S3] An S3 object configured to use the secrets bucket
```

File: lib/identity/hostdata.rb [hunk 4]
```diff
@@ -81,16 +81,31 @@ def self.in_datacenter
       yield self if in_datacenter?
     end
 
-    # @return [S3]
-    def self.s3(logger: default_logger, s3_client: nil)
+    # @return [S3] An S3 object onfigured to use the app-secrets bucket
```
Comment (zachmargolis) [inline lib/identity/hostdata.rb:84]: same typo
```suggestion
    # @return [S3] An S3 object configured to use the app-secrets bucket
```

File: lib/identity/hostdata/s3.rb (hunk 2 above)
Comment (jmhooper) [reply lib/identity/hostdata/s3.rb]: I made a shot at cleaning up some duplication in d5421dc1fc95d19424ac396ed3e5aa1e470d1574

File: lib/identity/hostdata/s3.rb [hunk 5]
```diff
@@ -16,33 +16,53 @@ def initialize(bucket:, env:, region:, s3_client: nil, logger: nil)
...
+
+      def stream_s3_object_to_file(s3_object_response, file_path)
+        FileUtils.mkdir_p(File.dirname(file_path))
+
+        File.open(file_path, 'wb') do |file|
+          while bytes = s3_object_response.body.read(512)
+            file.write(bytes)
+          end
+        end
+      end
```
Comment (zachmargolis) [inline lib/identity/hostdata/s3.rb:59]: This seems to me like the kind of thing we should not re-implement from the S3 gem

As we discussed in Slack, if we need to be able to remove FakeFS to inherit this behavior from the gem, let's remove FakeFS

File: spec/identity/hostdata/s3_spec.rb [hunk 6]
```diff
@@ -1,12 +1,13 @@
 require 'spec_helper'
 
 RSpec.describe Identity::Hostdata::S3 do
-  let(:fake_s3) { Identity::Hostdata::FakeS3Client.new }
-
   around(:each) do |ex|
     Identity::Hostdata.reset!
 
+    # set up before FakeFS
```
Comment (zachmargolis) [inline spec/identity/hostdata/s3_spec.rb:7]: this comment is a lie now... we no longer use fakefs

```suggestion
```

File: spec/identity/hostdata_spec.rb [hunk 7]
```diff
@@ -225,7 +225,7 @@
     end
   end
 
-  describe '.s3' do
+  describe 'app_secrets_s3' do
```
Comment (zachmargolis) [inline spec/identity/hostdata_spec.rb:228]: nit: use dot notation for class methods

```suggestion
  describe '.app_secrets_s3' do
```

File: spec/identity/hostdata_spec.rb [hunk 8]
```diff
@@ -258,7 +258,48 @@
...
-      subject(:s3) { Identity::Hostdata.s3(logger: logger) }
+      subject(:s3) { Identity::Hostdata.app_secrets_s3(logger: logger) }
+
+      it 'passes the logger through' do
+        expect(s3.logger).to eq(logger)
+      end
+    end
+  end
+
+  describe 'secrets_s3' do
```
Comment (zachmargolis) [inline spec/identity/hostdata_spec.rb:269]: ditto, dot notation

```suggestion
  describe '.secrets_s3' do
```

**Review by zachmargolis — APPROVED** [review]
LGTM! a few nits
